### PR TITLE
[WIP, Prototype] F# Compiler Server - Out of Process tooling for VS

### PR DESCRIPTION
**Pull request by TIHan**
**Highly work in progress**

This is the main PR for a prototype of a F# compiler server that will serve as an out of proc solution for tooling in Visual Studio. More information to come.

### CODE REVIEW

**Comment by cartermp** [comment]
A quick guide on this:

**Why**

The VS process is not sufficient for all of F# tooling for medium-to-large projects. What can happen in these solutions is that the tools consume too much memory over time, leading to consistent and often lengthy UI delays as the GC reclaims memory. But due to requests coming into the compiler service, this memory is quickly re-allocated, leading to constant UI delays until the user restarts VS. This experience is hardly acceptable.

**Solution**

The "best" solution is to "fix" the compiler service so that it doesn't allocate much, but doing so would likely involve a wholesale redesign of various components that would take a very long time before users could really see any benefits. It would likely involve far more churn and risk as well.

So, the next best solution is to give the tools more room to work with. Thus, moving the compiler service into a 64-bit process that VS calls into is the answer.

**Risks involved**

Despite FSharp.Compiler.Private being monolithic, and thus "easy" to move over, it's not a silver bullet:

* There is a nonzero cost to sending data to and from the process
* The process itself could begin to allocate significant amounts of memory depending on user actions
* As with any nontrivial change, there will likely be a bug tail involved

However, these risks are acceptable given the current unacceptable experience.

**Rollout**

Once all of FSharp.Editor's features call the server process, this will be ready to flight in a preview release of VS. It will be put behind a setting that you'll have to turn on. Once VS is restarted, tooling with use these process instead.

After validating through lots of internal dogfooding and users who are willing to try this on their solutions, we'll flip the switch and make this on by default for an update to VS 2019.

**Future use**

The the server process is a .NET Core application, so it will be cross-platform. If suitable for Ionide, it could potentially be used as a drop-in replacement for FSAC. We can figure out what to do when the time is appropriate.

**Comment by realvictorprm** [comment]
Cool stuff @TiHan!
Thank you very much for elaborating @cartermp :)

File: vsintegration/src/FSharp.Compiler.Server/Ipc/IpcServer.fs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+        let writer = new StreamWriter(fcs, AutoFlush = true)
+        let reader = new StreamReader(fcs)
+
+        try
+            while true do
+                if not fcs.IsConnected then
+                    failwith "Client disconnected"
+
+                let count = reader.Read(buffer, 0, buffer.Length)
```
Comment (saul) [inline vsintegration/src/FSharp.Compiler.Server/Ipc/IpcServer.fs]: Better off using ReadLine/WriteLine.

It's possible that:

1. The buffer is too small to hold the entire message (and so you end up reading a partial message)
2. There are multiple messages in the buffer (and so the JSON deserialisation fails)

You can tell Newtonsoft.Json to serialise a JSON object on a single-line by using Formatting.None.

File: vsintegration/src/FSharp.Compiler.Server/Ipc/IpcClient.fs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        Reply: AsyncReplyChannel<Result<'Receive, string>>
+    }
+
+type IpcMessageClient<'Send, 'Receive>(name) =
+
+    let buffer = Array.zeroCreate<char> Constants.IpcBufferSize
+
+    let restartingEvent = Event<unit>()
+
+    let mutable agent = Unchecked.defaultof<_>
```
Comment (saul) [inline vsintegration/src/FSharp.Compiler.Server/Ipc/IpcClient.fs]: Better to make invalid states impossible, and an IpcMessageClient is constructed in the started state.

File: vsintegration/src/FSharp.Compiler.Server/Behavior/GetErrorInfos.fs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+                    | FSharpCheckFileAnswer.Aborted -> return [||]
+                    | FSharpCheckFileAnswer.Succeeded results ->
+                        // In order to eleminate duplicates, we should not return parse errors here because they are returned by `AnalyzeSyntaxAsync` method.
+                        let allErrors = HashSet(results.Errors, errorInfoEqualityComparer)
+                        allErrors.ExceptWith(parseResults.Errors)
+                        return Seq.toArray allErrors
+                | GetErrorInfosCommandType.Syntax ->
+                    return parseResults.Errors
+                | _ -> 
+                    return failwith "should not happen"
```
Comment (saul) [inline vsintegration/src/FSharp.Compiler.Server/Behavior/GetErrorInfos.fs]: Not very informative :laughing:

**Comment by cartermp** [comment]
We wouldn't consider LSP support at least until LSP can support all of what we offer in the IDE.

**Comment by cloudRoutine** [comment]
@cartermp doesn't it make sense to support it at least as an interface to some of the functionality. It's not like the VS implementation has to rely on LSP, but by supporting the LSP [a bunch of other editors](https://langserver.org/#implementations-client) can get decent baseline level of F# language support with minimal client configuration.

**Currently LSP 3.0 supports -**
* completion
* completion resolve
* hover
* signatureHelp
* declaration
* definition
* typeDefinition
* implementation
* references
* documentHighlight
* documentSymbol
* codeAction
* codeLens
* codeLens resolve
* documentLink
* documentLink resolve
* documentColor
* colorPresentation
* formatting
* rangeFormatting
* onTypeFormatting
* rename
* prepareRename
* foldingRange

Anything beyond that can always be executed through custom commands, it's just a question of the amount of flexibility that the client for each particular editor has to integrate additional functionality.

**Comment by saul** [comment]
Completely agree with @cloudRoutine here. It seems like a huge missed opportunity to not use LSP here. We’re going to end up implementing a very similar but still distinct protocol - why not just start with LSP?

**Comment by realvictorprm** [comment]
Same opinion here, it's not a big deal to start with LSP and add custom commands.

**Comment by cartermp** [comment]
We can consider something like an LSP interface (that we will not use) once we get to the point of this being meaningful outside of VS. It's premature to plan for a far-off future when this prototype doesn't even support half of the VS F# features yet.

**Comment by TIHan** [comment]
Recently been having discussions on LSP and OOP with the Roslyn team. It seems the path forward to OOP is LSP for us. There is nothing concrete about any timeline to make this happen, but it is something that we want to do. 

While the prototype was interesting and proved that it can be done in the manner that I chose, it is not the right way to do OOP in VS.

**Pull request closed without merging**